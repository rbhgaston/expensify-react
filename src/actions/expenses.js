// import {v4 as uuidv4} from 'uuid';
import db from '../firebase/firebase'

// ADD
export const  addExpense = (expense) => ({
    type: 'ADD',
    expense
})

export const startAddExpense = ({description= '', note= '', createdAt= 0, amount= 0} = {}) =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        const expense = {
            description,
            amount,
            note,
            createdAt
        }
        return db.ref(`users/${uid}/expenses`).push(expense)
            .then(ref => dispatch(addExpense({
                id: ref.key,
                ...expense
            })))
    }

// EDIT
export const editExpense = (id, updates) => ({
    type: 'EDIT',
    id,
    updates
})

export const startEditExpense= (id, updates) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).update(updates)
            .then(() => dispatch(editExpense(id, updates)))
    }



// REMOVE
export const removeExpense = (id) => ({type: 'REMOVE', id})

export const startRemoveExpense = (id) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).remove()
            .then(() => dispatch(removeExpense(id)))
        }

// SET
export const setExpenses = (expenses) => ({
    type: 'SET',
    expenses
})

export const startSetExpenses = () =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses`).once('value').then(snapshot => {
            const expensesArray = []
            snapshot.forEach(expense => {expensesArray.push({id: expense.key.toString(), ...expense.val()})})
            dispatch(setExpenses(expensesArray))
        })
    }