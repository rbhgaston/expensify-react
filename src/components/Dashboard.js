import React from "react";
import ExpensesList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

export default function DashBoard() {
    return (
        <div>
            <ExpenseListFilters />
            <ExpensesList />
        </div>
    )
}