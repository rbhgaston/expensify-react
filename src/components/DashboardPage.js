import React from "react";
import ExpensesList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseListSummary from './ExpenseListSummary';

export default function DashboardPage() {
    return (
        <div>
            <ExpenseListSummary />
            <ExpenseListFilters />
            <ExpensesList />
        </div>
    )
}