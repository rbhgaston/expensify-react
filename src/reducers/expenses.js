const expensesReducerDefaultState = []
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD':
            return [
                ...state,
                action.expense
            ]
        case 'EDIT':
            return state.map(expense => expense.id === action.id ? {...expense, ...action.updates} : expense )
        case 'REMOVE':
            return state.filter(expense => expense.id !== action.id)
        default:
            return state
    }
}

export default expensesReducer