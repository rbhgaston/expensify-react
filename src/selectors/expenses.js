import moment from 'moment';

export const getFilteredExpenses =  (expenses, {text, sort, startDate, endDate}) => {
    return expenses.filter(expense => {
        const isGreaterThanStart = startDate ?  moment(expense.createdAt).isSameOrAfter(startDate, 'day') : true
        const isLessThanEnd = endDate ? moment(expense.createdAt).isSameOrBefore(endDate, 'day') : true

        return expense.description.includes(text) && isGreaterThanStart && isLessThanEnd
        })
        .sort((a, b) => sort === 'amount' ? a.amount - b.amount : a.createdAt - b.createdAt)
}

export const getTotalAmount = (expenses = []) => expenses.reduce((total, expense) => total + expense.amount, 0)