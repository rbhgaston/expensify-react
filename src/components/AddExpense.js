import React from 'react';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';
import {connect} from 'react-redux'

export function AddExpense(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.addExpense(expense)
                    props.history.push('/')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

// if you pass mapDispatchToProps, you lose access to dispatch in props
export default connect(undefined, mapDispatchToProps)(AddExpense)

