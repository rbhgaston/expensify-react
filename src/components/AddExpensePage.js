import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {startAddExpense, addExpense} from '../actions/expenses';

export function AddExpensePage(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.startAddExpense(expense)
                    props.history.push('/')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

// if you pass mapDispatchToProps, you lose access to dispatch in props
export default connect(undefined, mapDispatchToProps)(AddExpensePage)

