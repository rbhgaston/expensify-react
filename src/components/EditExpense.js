import React from "react";
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';
import {connect} from 'react-redux';

export class EditExpense extends React.Component{
    onSubmit= (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }

    onClick = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }

    render(){
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        )
    }
}


const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(({id}) => id === props.match.params.id)
})
// 2nd arg
const mapDispatchToProps = dispatch => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id))

})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense)