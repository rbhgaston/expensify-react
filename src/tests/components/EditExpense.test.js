import React from 'react';
import {EditExpense} from '../../components/EditExpense';
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses';

const expenseIndex = 0
let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;
beforeEach(() => {
    editExpenseSpy = jest.fn()
    removeExpenseSpy = jest.fn()
    historySpy = {push: jest.fn()}

    wrapper = shallow(
        <EditExpense
            editExpense={editExpenseSpy}
            removeExpense={removeExpenseSpy}
            history={historySpy}
            expense={expenses[expenseIndex]}
        />)
})

test('render EditExpense', () => {
    expect(wrapper).toMatchSnapshot()
})


test('call editExpense and history.push in onSubmit with right args', () => {
    wrapper.find('ExpenseForm').simulate('submit', expenses[expenseIndex])

    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenseIndex, expenses[expenseIndex])
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
})
// TODO simulate vs prop()()
test('call removeExpense and history.push in onClick with right args', () => {
    wrapper.find('button').simulate('click')

    expect(removeExpenseSpy).toHaveBeenLastCalledWith(expenseIndex)
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
})