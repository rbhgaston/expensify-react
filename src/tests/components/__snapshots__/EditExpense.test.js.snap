// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render EditExpense 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 100,
        "createdAt": 777600000,
        "description": "test0",
        "id": 0,
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
